### ISSUE TITLE
[Bug] unexpected lines_offsets when updates event

### ISSUE THREAD

**Issue by GiveMe-A-Name** [GiveMe-A-Name/lsp-textdocument#28]
follow below Examples: 

```rust
#[test]
fn test_line_offsets() {
    let mut text_document = full_text_document();
    text_document.update(
        &[TextDocumentContentChangeEvent {
            range: Some(Range {
                start: Position {
                    line: 0,
                    character: 0,
                },
                end: Position {
                    line: 0,
                    character: 0,
                },
            }),
            range_length: None,
            text: "0:1332533\n0:1332534\n0:1332535\n0:1332536\n".to_string(),
        }],
        1,
    );
    assert_eq!(text_document.line_offsets[0..4], vec!(0, 10, 20, 30));

    text_document.update(
        &[
            TextDocumentContentChangeEvent { range: Some(Range { start: Position { line: 1, character: 10 }, end: Position { line: 2, character: 10 } }), range_length: None, text: "1:6188912\n1:6188913\n1:6188914\n1:6188915\n1:6188916\n1:6188917\n1:6188918\n1:6188919\n1:6188920\n".to_string() },
        ],
        2
    );
    assert!(text_document
        .get_content(None)
        .to_string()
        .starts_with("0:1332533\n0:1332534\n1:6188912\n1:6188913\n"));
    // Apparently text_document.line_offsets should be [0, 10, 20, 30] ......
    // panic! 
    //   left: `[0, 10, 30, 40]`,
    //   right: `[0, 10, 20, 30]`'

    assert_eq!(text_document.line_offsets[0..4], vec!(0, 10, 20, 30));
}
```

There are some bugs in line_offsets update logic.

**Comment by GiveMe-A-Name** [comment]
Actually,  The  examples haven a invalid Position - `Position { line: 1, character: 10 } `

In the debug process, we can see the line_1 only include [10] - [19]

![image](https://github.com/GiveMe-A-Name/lsp-textdocument/assets/58852732/7a3fd1c9-25cd-498b-a764-9c124b902b22)

In original design, the language server believe all message that sent by client is valid.

So we haven't to solve this situations.

**Comment by bolinfest** [comment]
While this seems that this is what VS Code *sends*, I'm trying to find something in the spec that says using the index after a newline is invalid:

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position

That said, the thing that would probably unblock me is updating `FullTextDocument::position_at()` to use the start of the next line when `offset` corresponds to `\n`.

@mroch did you look at this at all when introducing `position_at()`?

**Comment by bolinfest** [comment]
I see that the Rust `position_at()` implementation mirrors the TypeScript one in VS Code:

https://github.com/microsoft/vscode-languageserver-node/blob/f2ff7d55464a1f58f978cb6635bd8865f050553c/textDocument/src/main.ts#L281-L301

Note that if you have a string of with N characters `"abc"`, there are N+1 possible cursor positions:

1. before `a`
2. between `a` and `b`
3. between `b` and `c`
4. after `c`

so a 3-character string has 4 possible cursor positions, so I'm not convinced by your screenshot that the specified `Position` is invalid.

**Comment by bolinfest** [comment]
I just ran the following to test out the equivalent TypeScript code:

```shell
$ mkdir /tmp/lsptest
$ cd /tmp/lsptest
$ yarn
$ yarn add vscode-languageserver-node
$ node
```

Then in the Node shell:

```
> var vs = require('vscode-languageserver')
> var doc = vs.TextDocument.create('', 'text', 1, "0:1332533\n0:1332534\n0:1332535\n0:1332536\n")
> event = {range: {start: {line: 1, character: 10}, end: {line: 2, character: 10}}, text: "1:6188912\n1:6188913\n1:6188914\n1:6188915\n1:6188916\n1:6188917\n1:6188918\n1:6188919\n1:6188920\n"}
> doc.update(event, 2)
> doc
FullTextDocument {
  _uri: '',
  _languageId: 'text',
  _version: 2,
  _content: '1:6188912\n' +
    '1:6188913\n' +
    '1:6188914\n' +
    '1:6188915\n' +
    '1:6188916\n' +
    '1:6188917\n' +
    '1:6188918\n' +
    '1:6188919\n' +
    '1:6188920\n',
  _lineOffsets: undefined
}
> doc.getLineOffsets()
[
   0, 10, 20, 30, 40,
  50, 60, 70, 80, 90
]
```

**Comment by bolinfest** [comment]
Hmm, though as @mroch pointed out to me via chat, this seems even stranger, since it looks like the entire document contents have been replaced even though `event` should be treated as an incremental update?

**Comment by bolinfest** [comment]
Weird, when I look at what is actually in my `node_modules` folder, I find this implementation for `update()`:

```
    FullTextDocument.prototype.update = function (event, version) {
        this._content = event.text;
        this._version = version;
        this._lineOffsets = undefined;
    };
```

Looking at `applyEdits()` instead now...

**Comment by bolinfest** [comment]
OK, I have a better repro. Run this script via `node` in a folder where you have run `yarn add vscode-languageserver-node`:

```javascript
var vs = require("vscode-languageserver");
var doc = vs.TextDocument.create(
  "",
  "text",
  1,
  "0:1332533\n0:1332534\n0:1332535\n0:1332536\n"
);
var event = {
  range: { start: { line: 1, character: 10 }, end: { line: 2, character: 10 } },
  text: "1:6188912\n1:6188913\n1:6188914\n1:6188915\n1:6188916\n1:6188917\n1:6188918\n1:6188919\n1:6188920\n",
};

// This does not work?
// doc.update([event], 2);

var textEdit = {
  range: event.range,
  newText: event.text,
};

var result = vs.TextDocument.applyEdits(doc, [textEdit]);
console.log(result);
```

I get:

```
0:1332533
0:1332534
1:6188912
1:6188913
1:6188914
1:6188915
1:6188916
1:6188917
1:6188918
1:6188919
1:6188920
0:1332536
```

**Comment by bolinfest** [comment]
Note also these docstrings:

https://github.com/microsoft/vscode-languageserver-node/blob/f2ff7d55464a1f58f978cb6635bd8865f050553c/textDocument/src/main.ts#L146-L174

**Comment by mroch** [comment]
@bolinfest TextDocument from vscode-languageserver is deprecated, try using vscode-languageserver-textdocument

```
var vs = require("vscode-languageserver-textdocument");
var doc = vs.TextDocument.create(
  "",
  "text",
  1,
  "0:1332533\n0:1332534\n0:1332535\n0:1332536\n"
);
var event = {
  range: { start: { line: 1, character: 10 }, end: { line: 2, character: 10 } },
  text: "1:6188912\n1:6188913\n1:6188914\n1:6188915\n1:6188916\n1:6188917\n1:6188918\n1:6188919\n1:6188920\n",
};

// now this works
vs.TextDocument.update(doc, [event], 2);

console.log(doc);
```

```
FullTextDocument {
  _uri: '',
  _languageId: 'text',
  _version: 2,
  _content: '0:1332533\n' +
    '0:1332534\n' +
    '1:6188912\n' +
    '1:6188913\n' +
    '1:6188914\n' +
    '1:6188915\n' +
    '1:6188916\n' +
    '1:6188917\n' +
    '1:6188918\n' +
    '1:6188919\n' +
    '1:6188920\n' +
    '0:1332536\n',
  _lineOffsets: [
      0,  10,  30, 40,  50,
     60,  70,  80, 90, 100,
    110, 110, 120
  ]
}
```

**Comment by mroch** [comment]
`lineOffsets: [0,  10,  30, ...]` ??

so, i think this is also broken upstream

**Comment by GiveMe-A-Name** [comment]
Fixed by #31
